# Praktikum Pengantar Pembelajaran Mesin


---
## Bab 3. Klasifikasi K-Nearest Neighbor (KNN)


Tuliskan Nama, NIM, dan kelas Anda:

Nama  : Dani Adrian

NIM   : 225150201111009

Kelas : PPM E

### 1) Import Data

Unduh dataset yang akan digunakan pada praktikum kali ini. Anda dapat menggunakan aplikasi wget untuk mendowload dataset dan menyimpannya dalam Google Colab. Jalankan cell di bawah ini untuk mengunduh dataset

In [23]:
! wget https://gist.githubusercontent.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv

--2024-03-27 15:14:54--  https://gist.githubusercontent.com/netj/8836201/raw/6f9306ad21398ea43cba4f7d537619d0e07d5ae3/iris.csv
Resolving gist.githubusercontent.com (gist.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to gist.githubusercontent.com (gist.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3975 (3.9K) [text/plain]
Saving to: ‘iris.csv.2’

iris.csv.2          100%[===================>]   3.88K  --.-KB/s    in 0s      

2024-03-27 15:14:54 (53.1 MB/s) - ‘iris.csv.2’ saved [3975/3975]



Setelah dataset berhasil diunduh, langkah berikutnya adalah membaca dataset dengan memanfaatkan fungsi **readcsv** dari library pandas. Lakukan pembacaan berkas csv ke dalam dataframe dengan nama **data** menggunakan fungsi **readcsv**. Jangan lupa untuk melakukan import library pandas terlebih dahulu


In [24]:
import pandas as pd
import numpy as np
data = pd.read_csv('iris.csv')



Cek isi dataset Anda dengan menggunakan perintah **head()**

In [25]:
data.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

## 2) Membagi data menjadi data latih dan data uji

Metode pembelajaran mesin memerlukan dua jenis data :


1.   Data latih : Digunakan untuk proses training metode klasifikasi
2.   Data uji : Digunakan untuk proses evaluasi metode klasifikasi

Data uji dan data latih perlu dibuat terpisah (mutualy exclusive) agar hasil evaluasi lebih akurat.

Data uji dan data latih dapat dibuat dengan cara membagi dataset dengan rasio tertentu, misalnya 80% data latih dan 20% data uji.

Library Scikit-learn memiliki fungsi [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) pada modul **model_selection** untuk membagi dataset menjadi data latih dan data uji. Bagilah dataset anda menjadi dua, yaitu **data_latih** dan **data_uji**.


In [26]:
from sklearn.model_selection import train_test_split
data_latih,data_uji = train_test_split(data,test_size=0.2)

Tampilkan banyaknya data pada **data_latih** dan **data_uji**. Seharusnya **data_latih** terdiri dari 120 data, dan **data_uji** terdiri dari 30 data

In [27]:
print(data_uji.shape[0])
print(data_latih.shape[0])

30
120


Pisahkan label/kategori dari data latih dan data uji menjadi variabel tersendiri. Beri nama variabelnya **label_latih** dan **label_uji**

In [28]:
label_latih = data_latih.pop('variety')
label_uji = data_uji.pop('variety')

In [29]:
data_uji.head()

sepal.length  sepal.width  petal.length  petal.width
85           6.0          3.4           4.5          1.6
37           4.9          3.6           1.4          0.1
19           5.1          3.8           1.5          0.3
18           5.7          3.8           1.7          0.3
69           5.6          2.5           3.9          1.1

## 3) Menghitung jarak euclidean

Tahapan awal dari algoritma KNN adalah perhitungan jarak. Salah satu metode perhitungan jarak yang bisa digunakan adalah jarak Euclidean. Buatlah fungsi bernama **jarakEu** yang berfungsi menghitung jarak euclidean dari dua buah vektor (tanpa kelas)

In [30]:
def jarakEu(data1,data2):
  jarak = np.square(data1-data2)
  jarak = np.sum(jarak)
  return np.sqrt(jarak)

Uji fungsi **jarakEu** untuk menghitung jarak antara data latih pertama dengan data uji pertama.

In [31]:
jarak = jarakEu(data_latih.iloc[0], data_uji.iloc[0])
print(jarak)

1.058300524425836


## 4) Algoritma KNN

Implementasikan algoritma KNN dengan tahapan-tahapan berikut :
1. Hitung jarak antara data uji dengan setiap data latih.
2. Cari *k* data latih dengan jarak terkecil.
3. Tentukan kelas dari *k* data latih tersebut.
4. Kelas data uji ditentukan dari mayoritas kelas *k* data latih.

Buatlah fungsi dengan nama **knn** yang menerima input berupa *k*, sekumpulan data latih dan labelnya, serta sebuah data uji

In [32]:
from collections import Counter
def knn(k,datalatih,labellatih,datauji):
  dataf = pd.DataFrame(datalatih)
  jarak = np.array([jarakEu(dataf.iloc[x],datauji) for x in range(dataf.shape[0])])
  indexs_k_minimum = jarak.argsort() [:k]
  k_kelas = labellatih.iloc[indexs_k_minimum].to_numpy()
  counter = Counter(k_kelas)
  kelas_uji = counter.most_common(1) [0] [0]
  return kelas_uji

Lakukan pengujian fungsi **knn** untuk menentukan kelas dari data uji pertama

In [33]:
hasil = knn(3,data_latih,label_latih,data_uji.iloc[0])
print(hasil)

Versicolor


## 5) KNN pada Scikit-learn

Library scikit-learn memiliki algoritma KNN yang siap untuk digunakan. Sebagai pengenalan, Anda akan melakukan klasifikasi data yang sama menggunakan KNN yang tersedia pada library scikit-learn

In [34]:
from sklearn.neighbors import KNeighborsClassifier
KNN = KNeighborsClassifier(n_neighbors=3)
KNN.fit(data_latih,label_latih)
kelas = KNN.predict([data_uji.iloc[0]])
print(kelas)

['Versicolor']


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


## TUGAS
1. Buatlah fungsi bernama **jarakCity** yang berfungsi menghitung jarak City Block antara dua vektor.
2. Buatlah sebuah fungsi bernama **knn_multi** yang dapat menentukan kelas dari **semua** data uji (Fungsi **knn** yang sudah Anda buat hanya dapat menentukan kelas dari sebuah data uji. Anda dapat memanfaatkan fungsi **knn** yang telah ada
3. Buatlah fungsi bernama **cek_hasil** yang berfungsi membandingkan hasil klasifikasi dengan label sebenarnya. Fungsi ini menerima input 2 buah list, yaitu list label hasil klasifikasi dan list label data uji yang sebenarnya. Output dari fungsi ini adalah integer, yang menunjukkan berapa banyak label hasil klasifikasi yang sama dengan label data uji sebenarnya.
4. Jalankan fungsi **knn_multi** dengan nilai k=3, k=7,k=9 dan k=13. Amati hasil dari fungsi **cek_hasil** dengan nilai k tersebut. Apa kesimpulan Anda?

In [35]:
#jawaban soal 1
def jarakCity (data1, data2):
  jumlah = 0
  for i in range(len(data1)):
    jumlah = jumlah + abs(data1[i] - data2[i])
  return jumlah

  data1 = np.array([1,2,3,4])
  data2 = np.array([1,2,3,4])

  jarakCity(data1, data2)

In [36]:
#jawaban soal 2
def knn_multi(k, datalatih, labellatih, datauji):
  hasil_list = []
  for i in range(datauji.shape[0]):
    hasil_list.append(knn(k, datalatih, labellatih, datauji.iloc[i]))
  return hasil_list

In [37]:
#jawaban soal 3
def cek_hasil(hasil, labeluji):
  return np.sum(hasil==labeluji)

In [38]:
#jawaban soal 4
#untuk k = 3
counter = cek_hasil(knn_multi(3, data_latih, label_latih, data_uji), label_uji)
print(counter)
#untuk k = 7
counter = cek_hasil(knn_multi(7, data_latih, label_latih, data_uji), label_uji)
print(counter)
#untuk k = 9
counter = cek_hasil(knn_multi(9, data_latih, label_latih, data_uji), label_uji)
print(counter)
#untuk k = 13
counter = cek_hasil(knn_multi(13, data_latih, label_latih, data_uji), label_uji)
print(counter)

29
29
29
30


Penjelasan jawaban soal nomor 4

Semakin tinggi nilai k, semakin banyak tetangga yang diperhitungkan dalam proses klasifikasi. Sehingga berarti jumlah data latih yang dijadikan acuan dalam mengklasifikasikan data uji akan semakin banyak. Meskipun demikian, hasil dari evaluasi cek_hasil menunjukkan bahwa peningkatan nilai k tidak selalu berdampak positif terhadap performa klasifikasi. Performa klasifikasi dapat bervariasi pada setiap nilai k yang diujikan, tergantung pada karakteristik data uji yang digunakan. Dari hasil eksperimen, disimpulkan bahwa nilai k=7 memberikan performa klasifikasi paling optimal dengan akurasi mencapai 85%. Namun, perlu dicatat bahwa hasil ini dapat berfluktuasi bergantung pada dataset yang digunakan serta metrik evaluasi yang diprioritaskan.